File: client/packages/uml-glsp/src-old/uml-types.ts
/********************************************************************************
 * Copyright (c) 2021-2022 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * https://www.eclipse.org/legal/epl-2.0, or the MIT License which is
 * available at https://opensource.org/licenses/MIT.
 *
 * SPDX-License-Identifier: EPL-2.0 OR MIT
 ********************************************************************************/
import { DefaultTypes } from '@eclipse-glsp/protocol';

export namespace UmlTypes {
    // OBJECT DIAGRAM
    export const OBJECT = `${DefaultTypes.NODE}:object`;
    export const LINK = `${DefaultTypes.EDGE}:link`;
    export const ATTRIBUTE = `${DefaultTypes.NODE}:attribute`;

    // ACTIVITY DIAGRAM
    export const ACTIVITY = `${DefaultTypes.NODE}:activity`;
    export const ICON_ACTIVITY = `${ICON}:activity`;
    export const PARTITION = `${DefaultTypes.NODE}:partition`;
    export const ACTION = `${DefaultTypes.NODE}:action`;
    export const ICON_ACTION = `${ICON}:action`;
    export const SENDSIGNAL = `${DefaultTypes.NODE}:sendsignal`;
    export const ACCEPTEVENT = `${DefaultTypes.NODE}:acceptevent`;
    export const TIMEEVENT = `${DefaultTypes.NODE}:timeevent`;
    export const CALL = `${DefaultTypes.NODE}:call`;
    export const CALL_REF = `${DefaultTypes.LABEL}:callref`;
    export const CONTROLFLOW = `${DefaultTypes.EDGE}:controlflow`;
    export const LABEL_FLOW_GUARD = `${DefaultTypes.LABEL}:guard`;
    export const LABEL_FLOW_WEIGHT = `${DefaultTypes.LABEL}:weight`;
    // ControlNodes
    export const INITIALNODE = `${DefaultTypes.NODE}:initialnode`;
    export const FINALNODE = `${DefaultTypes.NODE}:finalnode`;
    export const FLOWFINALNODE = `${DefaultTypes.NODE}:flowfinalnode`;
    export const DECISIONMERGENODE = `${DefaultTypes.NODE}:decisionnode`;
    export const FORKJOINNODE = `${DefaultTypes.NODE}:forknode`;
    export const PARAMETER = `${DefaultTypes.NODE}:parameter`;
    export const PIN = `${DefaultTypes.NODE}:pin`;
    export const PIN_PORT = `${DefaultTypes.NODE}:pinport`;
    export const CENTRALBUFFER = `${DefaultTypes.NODE}:centralbuffer`;
    export const DATASTORE = `${DefaultTypes.NODE}:datastore`;
    export const EXCEPTIONHANDLER = `${DefaultTypes.EDGE}:exceptionhandler`;
    export const INTERRUPTIBLEREGION = `${DefaultTypes.NODE}:interruptibleregion`;
    export const CONDITION = `${DefaultTypes.NODE}:condition`;
    export const COMMENT = `${DefaultTypes.NODE}:comment`;
    export const COMMENT_LINK = `${DefaultTypes.EDGE}:commentlink`;

    // USECASE DIAGRAM
    export const USECASE = `${DefaultTypes.NODE}:usecase`;
    export const ICON_USECASE = `${ICON}:usecase`;
    export const ACTOR = `${DefaultTypes.NODE}:actor`;
    export const ICON_ACTOR = `${ICON}:actor`;
    // export const PACKAGE = `${DefaultTypes.NODE}:package`;
    export const LABEL_PACKAGE_NAME = `${DefaultTypes.LABEL}:package:name`;
    // export const ICON_PACKAGE = `${ICON}:package`;
    export const COMPONENT = `${DefaultTypes.NODE}:component`;
    export const EXTENSIONPOINT = `${DefaultTypes.NODE}:extensionpoint`;
    export const EXTEND = `${DefaultTypes.EDGE}:extend`;
    export const INCLUDE = `${DefaultTypes.EDGE}:include`;
    export const GENERALIZATION = `${DefaultTypes.EDGE}:generalization`;
    export const CONNECTIONPOINT = `${DefaultTypes.EDGE}:connectionpoint`;
    export const USECASE_ASSOCIATION = `${DefaultTypes.EDGE}:usecase-association`;

    // DEPLOYMENT DIAGRAM
    export const ICON_DEPLOYMENT_NODE = `${ICON}:node`;
    export const DEPLOYMENT_NODE = `${DefaultTypes.NODE}:node`;
    export const ICON_ARTIFACT = `${ICON}:artifact`;
    export const ARTIFACT = `${DefaultTypes.NODE}:artifact`;
    export const COMMUNICATION_PATH = `${DefaultTypes.EDGE}:communicationpath`;
    export const DEPLOYMENT = `${DefaultTypes.EDGE}:deployment`;
    export const ICON_EXECUTION_ENVIRONMENT = `${ICON}:executionenvironment`;
    export const EXECUTION_ENVIRONMENT = `${DefaultTypes.NODE}:executionenvironment`;
    export const ICON_DEVICE = `${ICON}:device`;
    export const DEVICE = `${DefaultTypes.NODE}:device`;
    export const ICON_DEPLOYMENT_SPECIFICATION = `${ICON}:deploymentspecification`;
    export const DEPLOYMENT_SPECIFICATION = `${DefaultTypes.NODE}:deploymentspecification`;
    export const DEPLOYMENT_COMPONENT = `${DefaultTypes.NODE}:deploymentcomponent`;

    // STATE MACHINE DIAGRAM
    export const ICON_STATE_MACHINE = `${ICON}:state-machine`;
    export const STATE_MACHINE = `${DefaultTypes.NODE}:state-machine`;
    export const REGION = `${DefaultTypes.NODE}:region`;
    export const ICON_STATE = `${ICON}:state`;
    export const STATE = `${DefaultTypes.NODE}:state`;
    export const LABEL_VERTEX_NAME = `${DefaultTypes.LABEL}:vertex-name`;
    export const INITIAL_STATE = `${DefaultTypes.NODE}:initial-state`;
    export const DEEP_HISTORY = `${DefaultTypes.NODE}:deep-history`;
    export const SHALLOW_HISTORY = `${DefaultTypes.NODE}:shallow-history`;
    export const FORK = `${DefaultTypes.NODE}:fork`;
    export const JOIN = `${DefaultTypes.NODE}:join`;
    export const JUNCTION = `${DefaultTypes.NODE}:junction`;
    export const CHOICE = `${DefaultTypes.NODE}:choice`;
    export const ENTRY_POINT = `${DefaultTypes.NODE}:entry-point`;
    export const EXIT_POINT = `${DefaultTypes.NODE}:exit-point`;
    export const TERMINATE = `${DefaultTypes.NODE}:terminate`;
    export const FINAL_STATE = `${DefaultTypes.NODE}:final-state`;
    export const STATE_ENTRY_ACTIVITY = `${DefaultTypes.NODE}:state-entry-activity`;
    export const STATE_DO_ACTIVITY = `${DefaultTypes.NODE}:state-do-activity`;
    export const STATE_EXIT_ACTIVITY = `${DefaultTypes.NODE}:state-exit-activity`;
    export const TRANSITION = `${DefaultTypes.EDGE}:transition`;
    export const LABEL_TRANSITION_NAME = `${DefaultTypes.LABEL}:transition-name`;
    export const LABEL_TRANSITION_GUARD = `${DefaultTypes.LABEL}:transition-guard`;
    export const LABEL_TRANSITION_EFFECT = `${DefaultTypes.LABEL}:transition-effect`;
    export const LABEL_TRANSITION_TRIGGER = `${DefaultTypes.LABEL}:transition-trigger`;
}

File: server/app/src/main/java/com/borkdominik/big/glsp/uml/uml/UMLTypes.java
/********************************************************************************
 * Copyright (c) 2024 borkdominik and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * https://www.eclipse.org/legal/epl-2.0, or the MIT License which is
 * available at https://opensource.org/licenses/MIT.
 *
 * SPDX-License-Identifier: EPL-2.0 OR MIT
 ********************************************************************************/
package com.borkdominik.big.glsp.uml.uml;

import java.util.Collection;
import java.util.List;

import org.eclipse.emf.ecore.EObject;
import org.eclipse.uml2.uml.Abstraction;
import org.eclipse.uml2.uml.AcceptEventAction;
import org.eclipse.uml2.uml.Activity;
import org.eclipse.uml2.uml.ActivityFinalNode;
import org.eclipse.uml2.uml.ActivityNode;
import org.eclipse.uml2.uml.ActivityParameterNode;
import org.eclipse.uml2.uml.ActivityPartition;
import org.eclipse.uml2.uml.Actor;
import org.eclipse.uml2.uml.Artifact;
import org.eclipse.uml2.uml.Association;
import org.eclipse.uml2.uml.CentralBufferNode;
import org.eclipse.uml2.uml.CommunicationPath;
import org.eclipse.uml2.uml.Component;
import org.eclipse.uml2.uml.ControlFlow;
import org.eclipse.uml2.uml.DataType;
import org.eclipse.uml2.uml.DecisionNode;
import org.eclipse.uml2.uml.Dependency;
import org.eclipse.uml2.uml.Deployment;
import org.eclipse.uml2.uml.DeploymentSpecification;
import org.eclipse.uml2.uml.Device;
import org.eclipse.uml2.uml.ElementImport;
import org.eclipse.uml2.uml.Enumeration;
import org.eclipse.uml2.uml.EnumerationLiteral;
import org.eclipse.uml2.uml.ExecutionEnvironment;
import org.eclipse.uml2.uml.Extend;
import org.eclipse.uml2.uml.FinalState;
import org.eclipse.uml2.uml.FlowFinalNode;
import org.eclipse.uml2.uml.ForkNode;
import org.eclipse.uml2.uml.Generalization;
import org.eclipse.uml2.uml.Include;
import org.eclipse.uml2.uml.InformationFlow;
import org.eclipse.uml2.uml.InitialNode;
import org.eclipse.uml2.uml.InputPin;
import org.eclipse.uml2.uml.InstanceSpecification;
import org.eclipse.uml2.uml.Interaction;
import org.eclipse.uml2.uml.Interface;
import org.eclipse.uml2.uml.InterfaceRealization;
import org.eclipse.uml2.uml.JoinNode;
import org.eclipse.uml2.uml.Lifeline;
import org.eclipse.uml2.uml.LiteralBoolean;
import org.eclipse.uml2.uml.LiteralInteger;
import org.eclipse.uml2.uml.LiteralSpecification;
import org.eclipse.uml2.uml.LiteralString;
import org.eclipse.uml2.uml.Manifestation;
import org.eclipse.uml2.uml.MergeNode;
import org.eclipse.uml2.uml.Message;
import org.eclipse.uml2.uml.Model;
import org.eclipse.uml2.uml.Node;
import org.eclipse.uml2.uml.OpaqueAction;
import org.eclipse.uml2.uml.Operation;
import org.eclipse.uml2.uml.OutputPin;
import org.eclipse.uml2.uml.PackageImport;
import org.eclipse.uml2.uml.PackageMerge;
import org.eclipse.uml2.uml.Parameter;
import org.eclipse.uml2.uml.Pin;
import org.eclipse.uml2.uml.PrimitiveType;
import org.eclipse.uml2.uml.Property;
import org.eclipse.uml2.uml.Pseudostate;
import org.eclipse.uml2.uml.Realization;
import org.eclipse.uml2.uml.Region;
import org.eclipse.uml2.uml.SendSignalAction;
import org.eclipse.uml2.uml.Slot;
import org.eclipse.uml2.uml.State;
import org.eclipse.uml2.uml.StateMachine;
import org.eclipse.uml2.uml.Substitution;
import org.eclipse.uml2.uml.Transition;
import org.eclipse.uml2.uml.Usage;
import org.eclipse.uml2.uml.UseCase;

import com.borkdominik.big.glsp.server.core.model.BGTypeProvider;

public enum UMLTypes implements BGTypeProvider {
   ABSTRACT_CLASS("AbstractClass"),
   ABSTRACTION(List.of(Abstraction.class)),
   ACCEPT_EVENT_ACTION(List.of(AcceptEventAction.class)),
   ACTIVITY(List.of(Activity.class)),
   ACTIVITY_FINAL_NODE(List.of(ActivityFinalNode.class)),
   ACTIVITY_NODE(List.of(ActivityNode.class)),
   ACTIVITY_PARAMETER_NODE(List.of(ActivityParameterNode.class)),
   ACTIVITY_PARTITION(List.of(ActivityPartition.class)),
   ACTOR(List.of(Actor.class)),
   ACTOR_STICKFIGURE("ActorStickfigure"),
   AGGREGATION("Aggregation"),
   ARTIFACT(List.of(Artifact.class)),
   ASSOCIATION(List.of(Association.class)),
   CENTRAL_BUFFER_NODE(List.of(CentralBufferNode.class)),
   CHOICE("Choice"),
   CLASS(List.of(org.eclipse.uml2.uml.Class.class)),
   COMMUNICATION_PATH(List.of(CommunicationPath.class)),
   COMPONENT(List.of(Component.class)),
   COMPOSITION("Composition"),
   CONTROL_FLOW(List.of(ControlFlow.class)),
   DATA_TYPE(List.of(DataType.class)),
   DECISION_NODE(List.of(DecisionNode.class)),
   DEEP_HISTORY("DeepHistory"),
   DEPENDENCY(List.of(Dependency.class)),
   DEPLOYMENT(List.of(Deployment.class)),
   DEPLOYMENT_SPECIFICATION(List.of(DeploymentSpecification.class)),
   DEVICE(List.of(Device.class)),
   ELEMENT_IMPORT(List.of(ElementImport.class)),
   ENUMERATION(List.of(Enumeration.class)),
   ENUMERATION_LITERAL(List.of(EnumerationLiteral.class)),
   EXECUTION_ENVIRONMENT(List.of(ExecutionEnvironment.class)),
   EXTEND(List.of(Extend.class)),
   FINAL_STATE(List.of(FinalState.class)),
   FORK("Fork"),
   FORK_NODE(List.of(ForkNode.class)),
   FLOW_FINAL_NODE(List.of(FlowFinalNode.class)),
   GENERALIZATION(List.of(Generalization.class)),
   INCLUDE(List.of(Include.class)),
   INFORMATION_FLOW(List.of(InformationFlow.class)),
   INITIAL_NODE(List.of(InitialNode.class)),
   INITIAL_STATE("InitialState"),
   INPUT_PIN(List.of(InputPin.class)),
   INSTANCE_SPECIFICATION(List.of(InstanceSpecification.class)),
   INTERACTION(List.of(Interaction.class)),
   INTERFACE(List.of(Interface.class)),
   INTERFACE_REALIZATION(List.of(InterfaceRealization.class)),
   JOIN("Join"),
   JOIN_NODE(List.of(JoinNode.class)),
   LIFELINE(List.of(Lifeline.class)),
   LITERAL_BOOLEAN(List.of(LiteralBoolean.class)),
   LITERAL_INTEGER(List.of(LiteralInteger.class)),
   LITERAL_SPECIFICATION(List.of(LiteralSpecification.class)),
   LITERAL_STRING(List.of(LiteralString.class)),
   MANIFESTATION(List.of(Manifestation.class)),
   MERGE_NODE(List.of(MergeNode.class)),
   MESSAGE(List.of(Message.class)),
   MODEL(List.of(Model.class)),
   NODE(List.of(Node.class)),
   OPAQUE_ACTION(List.of(OpaqueAction.class)),
   OPERATION(List.of(Operation.class)),
   OUTPUT_PIN(List.of(OutputPin.class)),
   PACKAGE(List.of(org.eclipse.uml2.uml.Package.class)),
   PACKAGE_IMPORT(List.of(PackageImport.class)),
   PACKAGE_MERGE(List.of(PackageMerge.class)),
   PARAMETER(List.of(Parameter.class)),
   PIN(List.of(Pin.class)),
   PRIMITIVE_TYPE(List.of(PrimitiveType.class)),
   PROPERTY(List.of(Property.class)),
   PROPERTY_MULTIPLICITY("PropertyMultiplicity"),
   PROPERTY_TYPE("PropertyType"),
   PSEUDOSTATE(List.of(Pseudostate.class)),
   REALIZATION(List.of(Realization.class)),
   REGION(List.of(Region.class)),
   SEND_SIGNAL_ACTION(List.of(SendSignalAction.class)),
   SHALLOW_HISTORY("ShallowHistory"),
   SLOT(List.of(Slot.class)),
   STATE(List.of(State.class)),
   STATE_MACHINE(List.of(StateMachine.class)),
   SUBSTITUTION(List.of(Substitution.class)),
   TRANSITION(List.of(Transition.class)),
   USAGE(List.of(Usage.class)),
   USE_CASE(List.of(UseCase.class));

   private final String typeId;
   private final Collection<Class<? extends EObject>> handledElements;

   UMLTypes(final String typeId) {
      this(typeId, List.of());
   }

   UMLTypes(final List<Class<? extends EObject>> handledElements) {
      this(handledElements.get(0).getSimpleName(), handledElements);
   }

   UMLTypes(final String typeId, final List<Class<? extends EObject>> handledElements) {
      this.typeId = typeId;
      this.handledElements = handledElements;
   }

   @Override
   public String typeId() {
      return typeId;
   }

   @Override
   public Collection<Class<? extends EObject>> handledElements() {
      return handledElements;
   }

}
